compiler: correctly generate code for local named types

aykevl opened the pull request:
It is possible to create function-local named types:

    func foo() any {
        type named int
        return named(0)
    }

This patch makes sure they don't alias with named types declared at the package scope.

Bug originally found by Damian Gryski while working on reflect support.

aykevl committed:
compiler: correctly generate code for local named types

It is possible to create function-local named types:

    func foo() any {
        type named int
        return named(0)
    }

This patch makes sure they don't alias with named types declared at the
package scope.

Bug originally found by Damian Gryski while working on reflect support.

dgryski commented:
Seems to work.  I'll give it a better review in a bit.

Even passes this test:
```
package main

import "reflect"

func main() {

        for i := 0; i < 4; i++ {
                a := named(i)
                r := reflect.TypeOf(a)
                println(i, r.Kind().String())
        }
}

func named(n int) any {

        if n == 0 {
                type named bool
                return named(true)
        }

        if n == 1 {
                type named byte
                return named(2)
        }

        if n == 3 {
                type named uint16
                return named(3)
        }

        type named int
        return named(4)
}
```

dgryski commented:
Oddly the failing XML test doesn't pass with this patch.  So looks like there's something else we need to hunt down.  I'll try to figure out if it's this PR or something else is still broken.

dgryski commented:
Here's the extracted xml test that is still getting the wrong type info for the structs:
```
// Copyright 2011 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import (
        "encoding/xml"
        "log"
)

func main() {

        testCases := []struct {
                f    func() ([]byte, error)
                want string
                ok   bool
        }{
                {encodeXMLNS1, `<Test xmlns="http://example.com/ns"><Body>hello world</Body></Test>`, true},
                {encodeXMLNS2, `<Test><body xmlns="http://example.com/ns">hello world</body></Test>`, true},
                {encodeXMLNS3, `<Test xmlns="http://example.com/ns"><Body>hello world</Body></Test>`, true},
                {encodeXMLNS4, `<Test xmlns="http://example.com/ns"><Body>hello world</Body></Test>`, false},
        }

        for i, tc := range testCases {
                println(i, tc.want)
                if b, err := tc.f(); err == nil {
                        if got, want := string(b), tc.want; got != want {
                                log.Fatalf("%d: got %s, want %s \n", i, got, want)
                        }
                } else {
                        log.Fatalf("%d: marshal failed with %s", i, err)
                }
        }
}

func encodeXMLNS1() ([]byte, error) {

        type T struct {
                XMLName xml.Name `xml:"Test"`
                Ns      string   `xml:"xmlns,attr"`
                Body    string
        }

        s := &T{Ns: "http://example.com/ns", Body: "hello world"}
        return xml.Marshal(s)
}

func encodeXMLNS2() ([]byte, error) {

        type Test struct {
                Body string `xml:"http://example.com/ns body"`
        }

        s := &Test{Body: "hello world"}
        return xml.Marshal(s)
}

func encodeXMLNS3() ([]byte, error) {

        type Test struct {
                XMLName xml.Name `xml:"http://example.com/ns Test"`
                Body    string
        }

        //s := &Test{XMLName: Name{"http://example.com/ns",""}, Body: "hello world"} is unusable as the "-" is missing
        // as documentation states
        s := &Test{Body: "hello world"}
        return xml.Marshal(s)
}

func encodeXMLNS4() ([]byte, error) {

        type Test struct {
                Ns   string `xml:"xmlns,attr"`
                Body string
        }

        s := &Test{Ns: "http://example.com/ns", Body: "hello world"}
        return xml.Marshal(s)
}
```

Doubt this runs, but I can maybe try to extract the pieces so it's just reflect calls grabbing type info.

aykevl commented:
> but I can maybe try to extract the pieces so it's just reflect calls grabbing type info.

That would certainly be very useful for debugging.

dgryski commented:
Better test case:
```
package main

import (
	"reflect"
)

type xmlt struct{}

var xml xmlt

type xmlName string

func (xmlt) Marshal(a any) (string, error) {
	t := reflect.TypeOf(a)
	return dumpType(t), nil
}

func dumpType(t reflect.Type) string {
	if t.Kind() == reflect.Struct {
		s := "struct {"
		numField := t.NumField()
		for i := 0; i < numField; i++ {
			f := t.Field(i)
			s += " " + f.Name + " " + dumpType(f.Type)
			// every field except the last needs a semicolon
			if i < numField-1 {
				s += ";"
			}
		}
		s += " }"
		return s
	}
	if t.Kind() == reflect.Pointer {
		return "*" + dumpType(t.Elem())
	}
	return t.Kind().String()
}

func main() {

	testCases := []struct {
		f func() (string, error)
	}{
		{encodeXMLNS1},
		{encodeXMLNS2},
		{encodeXMLNS3},
		{encodeXMLNS4},
	}

	for i, tc := range testCases {
		println(i)
		if b, err := tc.f(); err == nil {
			println(string(b))
		}
	}
}

func encodeXMLNS1() (string, error) {

	type T struct {
		XMLName xmlName `xml:"Test"`
		Ns      string  `xml:"xmlns,attr"`
		Body    string
	}

	s := &T{Ns: "http://example.com/ns", Body: "hello world"}
	return xml.Marshal(s)
}

func encodeXMLNS2() (string, error) {

	type Test struct {
		Body string `xml:"http://example.com/ns body"`
	}

	s := &Test{Body: "hello world"}
	return xml.Marshal(s)
}

func encodeXMLNS3() (string, error) {

	type Test struct {
		XMLName xmlName `xml:"http://example.com/ns Test"`
		Body    string
	}

	//s := &Test{XMLName: Name{"http://example.com/ns",""}, Body: "hello world"} is unusable as the "-" is missing
	// as documentation states
	s := &Test{Body: "hello world"}
	return xml.Marshal(s)
}

func encodeXMLNS4() (string, error) {

	type Test struct {
		Ns   string `xml:"xmlns,attr"`
		Body string
	}

	s := &Test{Ns: "http://example.com/ns", Body: "hello world"}
	return xml.Marshal(s)
}
```

Output:
```
~/go/src/github.com/dgryski/bug/refxml $ tinygo run  main.go
0
*struct { XMLName string; Ns string; Body string }
1
*struct { Body string }
2
*struct { Body string }
3
*struct { Body string }
~/go/src/github.com/dgryski/bug/refxml $ go run  main.go
0
*struct { XMLName string; Ns string; Body string }
1
*struct { Body string }
2
*struct { XMLName string; Body string }
3
*struct { Ns string; Body string }
```

aykevl commented:
Much smaller reproducer:

```go
package main

import "reflect"

func main() {
        itf := test2()
        t := reflect.TypeOf(itf).Elem()
        println("numField:", t.NumField())
}

func test1() interface{} {
        type Test struct {
                B int
        }
        return &Test{}
}

func test2() interface{} {
        type Test struct {
                A int
                B int
        }
        return &Test{}
}
```

Prints 2 in `go run`, but 1 in `tinygo run`.

aykevl commented:
This makes it very clear that the wrong type is stored in the interface:

```go
package main

import "reflect"

func main() {
        itf := test2()
        kind := reflect.TypeOf(itf).Elem().Kind()
        println("kind:", kind.String())
}

func test1() interface{} {
        type Test int
        return new(Test)
}

func test2() interface{} {
        type Test byte
        return new(Test)
}
```

```
~/src/tinygo/tinygo$ go run tmp/issue3566.go
kind: uint8

~/src/tinygo/tinygo$ tinygo run tmp/issue3566.go
kind: int
```

aykevl commented:
Pretty sure I know what's wrong now, I'll update this PR with a fix.
In summary, while the PR fixes plain named types, it is in a very similar way buggy for pointers to local named types.

aykevl commented:
Should be fixed now (with an additional test case).

dgryski reviewed (approved):
LGTM.

deadprogram commented:
This PR does not appear to pass CircleCI tests of backward compatibility:
https://app.circleci.com/pipelines/github/tinygo-org/tinygo/9219/workflows/0510a406-46e8-4cb3-b234-868850b82027/jobs/40141?invite=true#step-118-8

deadprogram commented on compiler/interface.go:
```diff
@@ -386,57 +408,83 @@ var basicTypeNames = [...]string{
 	switch t := t.(type) {
 	case *types.Named:
-		return "named:" + t.String()
+		if t.Obj().Parent() != t.Obj().Pkg().Scope() {
```
It seems like in Go1.18/LLVM 14 that `t.Obj().Pkg()` is sometimes `nil` here, causing

`panic: runtime error: invalid memory address or nil pointer dereference`

See https://app.circleci.com/pipelines/github/tinygo-org/tinygo/9219/workflows/c192f68a-0c3e-4eb8-a10f-0c1fdb9aa64e/jobs/40162?invite=true#step-118-8

aykevl replied:
Probably not LLVM, but it could be Go 1.18. I'd have to take a closer look.

aykevl commented:
Indeed, this was an issue with Go 1.18. It should be fixed now.

deadprogram commented:
Looks like a number of other CI fails now, however :smiley_cat:

dgryski commented:
Need to rerun `go test -update`.

deadprogram commented:
@aykevl can you do that please?

aykevl commented:
> Need to rerun `go test -update`.

No, this was actually a regression after I fixed Go 1.18 support. I've fixed the fix, so that `go test ./compiler` now works.

dgryski reviewed (approved):
LGTM

dgryski commented:
Only CI fail was the circuitplay-express.

deadprogram commented:
> Only CI fail was the circuitplay-express.

Restarted it and passed.

Now merging, thanks @aykevl and @dgryski
